### PR TITLE
Make input size/channels optional

### PR DESCRIPTION
**Pull request by mitmul**
This PR is for https://github.com/pfnet/chainer/issues/2045

- [x] Linear
- [x] Convolution2D
- [x] Deconvolution2D
- [x] DilatedConvolution2D
- [x] StatelessLSTM
- [x] LSTM
- [x] MLPConvolution2D

### CODE REVIEW

**Comment by gwtnb** [comment]
s/ommitted/omitted/g ?

**Comment by mitmul** [comment]
@gwtnb Thanks. Fixed.

**Comment by delta2323** [comment]
Sorry for making you wait so long. As we update APIs of convolution and linear links (#2185), could you resolve the conflict? Also, we make a standardized way of handling uninitialized variables (#2072), could you update the implementation to use it?

File: chainer/links/connection/dilated_convolution_2d.py [hunk 1]
```diff
@@ -42,12 +43,62 @@ class DilatedConvolution2D(link.Link):
         W (~chainer.Variable): Weight parameter.
         b (~chainer.Variable): Bias parameter.
 
+    .. admonition:: Example
+
+        There are several ways to make a DilatedConvolution2D link.
+
+        Let an input vector ``x`` be:
+
+        >>> x = np.arange(1 * 3 * 10 * 10).astype('f').reshape(1, 3, 10, 10)
```
Comment (delta2323) [inline chainer/links/connection/dilated_convolution_2d.py]: ditto

**Comment by mitmul** [comment]
@delta2323 Thank you for reviewing, I fixed some points.

File: chainer/links/connection/deconvolution_2d.py [hunk 2]
```diff
@@ -63,12 +64,69 @@ class Deconvolution2D(link.Link):
...
+
+            In this case, all the other arguments are set to the default
+            values.
+
+            >>> l = L.Deconvolution2D(3, 7, 4)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 13, 13)
+
+        2. Omit ``in_channels``:
```
Comment (delta2323) [inline chainer/links/connection/deconvolution_2d.py]: ditto

File: chainer/links/connection/deconvolution_2d.py [hunk 3]
```diff
@@ -63,12 +64,69 @@ class Deconvolution2D(link.Link):
...
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 13, 13)
+
+            >>> l = L.Deconvolution2D(None, 7, 4)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 13, 13)
+
+        3. Give arguments other than the first three arguments besides omitting
```
Comment (delta2323) [inline chainer/links/connection/deconvolution_2d.py]: ditto

File: chainer/links/connection/dilated_convolution_2d.py [hunk 4]
```diff
@@ -46,12 +47,61 @@ class DilatedConvolution2D(link.Link):
...
+        >>> x = np.arange(1 * 3 * 10 * 10, dtype='f').reshape(1, 3, 10, 10)
+
+        1. Give the first three arguments explicitly:
+
+            >>> l = L.DilatedConvolution2D(3, 7, 5)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 6, 6)
+
+        2. Omit ``in_channels``:
```
Comment (delta2323) [inline chainer/links/connection/dilated_convolution_2d.py]: ditto

File: chainer/links/connection/dilated_convolution_2d.py [hunk 5]
```diff
@@ -46,12 +47,61 @@ class DilatedConvolution2D(link.Link):
...
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 6, 6)
+
+            >>> l = L.DilatedConvolution2D(None, 7, 5)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 7, 6, 6)
+
+        3. Give arguments other than the first three arguments besides omitting
```
Comment (delta2323) [inline chainer/links/connection/dilated_convolution_2d.py]: ditto

File: chainer/links/connection/linear.py [hunk 6]
```diff
@@ -42,12 +42,63 @@ class Linear(link.Link):
...
+        1. Give the first two arguments explicitly:
+
+            Those numbers are considered as the input size and the output size.
+
+            >>> l = L.Linear(5, 10)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 10)
+
+        2. Omit ``in_size`` (give the output size only as the first argument):
```
Comment (delta2323) [inline chainer/links/connection/linear.py]: ditto

File: chainer/links/connection/linear.py [hunk 7]
```diff
@@ -42,12 +42,63 @@ class Linear(link.Link):
...
+            >>> y = l(x)
+            >>> y.shape
+            (1, 10)
+
+            >>> l = L.Linear(None, 10)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 10)
+
+        3. Give arguments other than the first three arguments besides omitting
```
Comment (delta2323) [inline chainer/links/connection/linear.py]: ditto

File: chainer/links/connection/lstm.py [hunk 8]
```diff
@@ -61,15 +65,53 @@ class StatelessLSTM(LSTMBase):
...
+        1. Give both ``in_size`` and ``out_size`` arguments:
+
+            >>> l = L.StatelessLSTM(10, 20)
+            >>> c_new, h_new = l(c, h, x)
+            >>> c_new.shape
+            (1, 20)
+            >>> h_new.shape
+            (1, 20)
+
+        2. Omit ``in_size`` argument:
```
Comment (delta2323) [inline chainer/links/connection/lstm.py]: ditto

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 9]
```diff
@@ -116,9 +116,7 @@ def test_pickling_gpu(self):
 class TestDilatedConvolution2DParameterShapePlaceholder(unittest.TestCase):
 
     def setUp(self):
-        in_channels = None
-        self.link = links.DilatedConvolution2D(
-            in_channels, 2, 3, stride=2, pad=2, dilate=2)
+        self.link = links.DilatedConvolution2D(2, 3, stride=2, pad=2, dilate=2)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: Please restore `in_chennal=None` case.

**Comment by mitmul** [comment]
@delta2323 Thank you for reviewing! I fixed docs and the test.

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 10]
```diff
@@ -113,12 +113,16 @@ def test_pickling_gpu(self):
         self.check_pickling(cuda.to_gpu(self.x))
 
 
+@testing.parameterize(
+    {'init_args': {'args': (2, 3),
+                   'kwargs': {'stride': 2, 'pad': 2, 'dilate': 2}}},
+    {'init_args': {'args': (None, 2, 3),
+                   'kwargs': {'stride': 2, 'pad': 2, 'dilate': 2}}})
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: The followings are simpler.

```
@testing.parameterize(
{'args': (2, 3), 'kwargs': ...},
{'args': (None, 2, 3), 'kwargs': ...}
)
```

Comment (mitmul) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: Thanks. Fixed it :+1:

**Comment by mitmul** [comment]
Thanks for checking. I fixed wrong parts in the docstring of dilated_convolution_2d.

File: chainer/links/connection/linear.py [hunk 11]
```diff
@@ -39,12 +39,61 @@ class Linear(link.Link):
...
+            >>> y.shape
+            (1, 10)
+
+            >>> l = L.Linear(None, 10)
+            >>> y = l(x)
+            >>> y.shape
+            (1, 10)
+
+            When you omit the first argument, you need to specify the other
+            subsequent arguments from ``bias`` as keyword arguments. So the
```
Comment (delta2323) [inline chainer/links/connection/linear.py]: `bias` -> `nobias`

**Comment by mitmul** [comment]
@delta2323 Thank you for checking. I fixed the typo. Could you check this PR again?

**Pull request merged**